### ISSUE TITLE
Patterns for dispatch on subcomponents

### ISSUE THREAD

**Issue by simonbyrne** [CliMA/ClimateMachine.jl#1832]
We often use a pattern such as:
```julia
f(atmos::AtmosModel, args...) = f(atmos.component, atmos, args...)
f(component::ComponentTypeA, atmos, args...) = # do something
f(component::ComponentTypeB, atmos, args...) = # do something else
```
The original reason for this is that it avoided the explosion of type parameters when trying to do it directly:
```julia
f(atmos::AtmosModel{C1, C2, ..., ComponentTypeA}, args...) where {C1, C2, ...} = # do something
f(atmos::AtmosModel{C1, C2, ..., ComponentTypeB}, args...) where {C1, C2, ...} = # do something else
```
(and remembering which type parameter goes where).

One problem @kpamnany pointed out was that this only lets you dispatch on a single item.

An alternative pattern would be to define the signature via a function
```julia
f(atmos::p(AtmosModel, C=ComponentTypeA), args...) = # do something
f(atmos::p(AtmosModel, C=ComponentTypeB), args...) = # do something else
```
where `p(AtmosModel, C=XXX)` returns an appropriately parametrized type, e.g. something like
```julia
function p(T::DataType; params...)
    if !isempty(params)
        error("Invalid parameters $params")
    end
    return T
end

function p(T::UnionAll; params...)
    if haskey(params, T.var.name)
        rparams = Base.structdiff(values(params), NamedTuple{(T.var.name,)})
        return p(T{params[T.var.name]}; rparams...)
    else
        return (p(T{P}; params...) where {P})
    end
end
```
which gives
```julia
julia> struct Foo{A,B}
           a::A
           b::B
       end

julia> p(Foo, A=Float32)
Foo{Float32,P} where P

julia> p(Foo, B=Float32)
Foo{P,Float32} where P

julia> p(Foo)
Foo{P,P1} where P1 where P

julia> p(Foo, B=Float32,A=Int)
Foo{Int64,Float32}
```

An alternative would be to match via subtypes instead of name, e.g.
```julia
function psub(T::DataType, ts...)
    if !isempty(ts)
        error("Invalid parameters $params")
    end
    return T
end

function psub(T::UnionAll, ts...)
    for (i,t) in enumerate(ts)
        if t <: T.var.ub
            rts = (ts[1:i-1]..., ts[i+1:end]...)
            return psub(T{t}, rts...)
        end
    end
    return (psub(T{P}, ts...) where {P})
end
```
which gives
```julia
julia> struct Bar{A<:Integer,B<:AbstractFloat}
           a::A
           b::B
       end

julia> psub(Bar,Int)
Bar{Int64,P} where P

julia> psub(Bar,Float32)
Bar{P,Float32} where P
```
this means that we don't need to worry about the type parameter names, but it does mean that all parameters would need a unique upperbound (if you don't supply one, it will be `Any`).

**Comment by kpamnany** [comment]
@simonbyrne: I assume we can accommodate matching out `FT` in this? As required below...

In general, I like the second option which allows us to do something like:

```julia
function flux_first_order!(
    atmos::psub(AtmosModel{FT}, EDMF),
    flux::Grad,
    state::Vars,
    aux::Vars,
    t::Real,
    ts,
    direction,
) where {FT}
```

Or:
```julia
function f(atmos::psub(AtmosModel, EquilMoist, HydrostaticState))
```
And have `f` be called when the `AtmosModel` has been created with an `EquilMoist` moisture model _and_ a `HydrostaticState` reference state.

Cc: @charleskawczynski, @blallen, @jkozdon

**Comment by jkozdon** [comment]
Either way seems fine to me. If asked I like the first option, since I don't need to worry about the unique upper bound (which from my quick naive thoughts seems like it could be a pain) and feels more like a keyword argument to the type.

**Comment by kpamnany** [comment]
I'm okay with the first approach too; the only downside is you have to remember the type parameter names used (to say `M = EquilMoist` or `RS = HydrostaticState`), but that's not a big problem.

**Comment by charleskawczynski** [comment]
I've been thinking a lot about this since adding in the new tendency specification. I almost wonder if the gpu perf issue we had with keywords might be solved on 1.6 ([here](https://github.com/JuliaLang/julia/blob/master/NEWS.md#compilerruntime-improvements)).

The other thing I was thinking about was using dispatch for the first few arguments (which we'll need in any case), and then constructing / passing a NamedTuple from the top layer for the remaining arguments. This way we'd still have a fixed number of args. The NamedTuple could even be constructed based on submodels. I haven't had a chance to prototype what I'm thinking, but I think it could work.

**Comment by simonbyrne** [comment]
> I'm okay with the first approach too; the only downside is you have to remember the type parameter names used (to say `M = EquilMoist` or `RS = HydrostaticState`), but that's not a big problem.

We could give the parameters more verbose names (there's no reason they need to be single letters or acronyms).

**Comment by kpamnany** [comment]
> We could give the parameters more verbose names

Yes, but the downside is that this becomes more verbose... win some lose some. Either way, I'm good with it and think we should do it.

**Comment by blallen** [comment]
I like the idea of the `p()` function with keyword arguments, seems to give us everything we need.

**Comment by charleskawczynski** [comment]
One thing that I like about the NamedTuple approach is that we can more easily unit test each source/flux function. For example, in:

```julia
function flux(::Advect{Energy}, m, state, aux, t, ts, direction)
    return (state.ρu / state.ρ) * state.ρe
end
```
We don't need `aux, t, ts, or direction` to compute this flux. If we took a NamedTuple approach:
```julia
function flux(::Advect{Energy}, args::NamedTuple)
    return (args.state.ρu / args.state.ρ) * args.state.ρe
end
```
Then we could test `flux` by passing in a `NamedTuple` with the required structure. We might want to add a third argument (e.g., `precomputed::NamedTuple`) so that we can store additional pieces for things we'd like to precompute. I think the other approaches discussed glue us a bit more to the model structure. These methods are already defined (or can be defined) atomically, so dispatching (at least for `flux` and `source`) on a single type seems reasonable. As for the other methods (e.g., `compute_gradient_argument!`), other options may be better.
